fdbserver/EncryptKeyProxy.actor.cpp

```diff
@@ -43,43 +43,95 @@
 #include "flow/network.h"
 
 #include <boost/mpl/not.hpp>
+#include <limits>
 #include <string>
 #include <utility>
 #include <memory>
 
 #include "flow/actorcompiler.h" // This must be the last #include.
 
 namespace {
+
+struct CipherKeyValidityTS {
+	int64_t refreshAtTS;
+	int64_t expAtTS;
+};
+
 bool canReplyWith(Error e) {
 	switch (e.code()) {
 	case error_code_encrypt_key_not_found:
+	// FDB <-> KMS connection may be observing transient issues
+	// Caller processes should consider reusing 'non-revocable' CipherKeys iff ONLY below error codes lead to CipherKey
+	// refresh failure
+	case error_code_timed_out:
+	case error_code_connection_failed:
 		return true;
 	default:
 		return false;
 	}
 }
 
+CipherKeyValidityTS getCipherKeyValidityTS(Optional<int64_t> refreshInterval, Optional<int64_t> expiryInterval) {
+	int64_t currTS = (int64_t)now();
+	int64_t defaultTTL = FLOW_KNOBS->ENCRYPT_CIPHER_KEY_CACHE_TTL;
+
+	CipherKeyValidityTS validityTS;
+	validityTS.refreshAtTS =
+	    refreshInterval.present() && refreshInterval.get() > 0 ? currTS + refreshInterval.get() : currTS + defaultTTL;
+	if (expiryInterval.present()) {
+		if (expiryInterval.get() < 0) {
+			// Non-revocable CipherKey
+			validityTS.expAtTS = std::numeric_limits<int64_t>::max();
+		} else if (expiryInterval.get() > 0) {
+			validityTS.expAtTS = currTS + expiryInterval.get();
+		} else {
+			ASSERT(expiryInterval.get() == 0);
+			// None supplied, match expiry to refresh timestamp
+			validityTS.expAtTS = validityTS.refreshAtTS;
```

Comment:
For a revokable key, sounds like it is valid have a refershTS, but not an expireTS? In that case setting expireTS:=refreshTS looks confusing.

Reply:
Very interesting point. 
It is true that `refreshTS` and `expireTS` is same for Revocable keys, but thinking keeping them separate may make sense, for instance: though CiperhKey expire in 1 hours, one may want to refresh-keys every 45 mins, to avoid delays in fetching newer keys after expiry (KMS delays generating/rotating new keys for some reason). 

By default refresh == expiry, as caller didn't specify the behavior, otherwise, in a perfect world both Revocable and Non-Revocable keys can have `refreshInterval` & `expiryInterval` set. 
Does this make sense?

Reply:
Thanks for the explanation. I was confused about the intent of revocable key.